src/Agent.Listener/CapabilitiesProvider.cs

```diff
@@ -0,0 +1,291 @@
+                new Capability( "maven", "mvn" ),
+                new Capability( "MSBuild", "xbuild" ),
+                new Capability( "node.js", "node" ),
+                new Capability( "node.js", "nodejs" ),
+                new Capability( "npm" ),
+                new Capability( "python" ),
+                new Capability( "python3" ),
+                new Capability( "sh" ),
+                new Capability( "subversion", "svn" ),
+                new Capability( "ruby" ),
+                new Capability( "rake" ),
+                new Capability( "Xamarin.iOS", "mdtool", new string[] { "/Applications/Xamarin Studio.app/Contents/MacOS/mdtool" } ),
+                new Capability( "Xamarin.Android", "mandroid", new string[] { "/Library/Frameworks/Xamarin.Android.framework/Commands/mandroid" } )
+            };
+
+        private static readonly List<ToolCapability> _toolCapabilities =
+            new List<ToolCapability>
+            {
+                new ToolCapability( "xcode", "xcode-select", "-p" ),
+            };
+
+        private static readonly string[] _ignoredEnvVariables = new string[] {
+            "TERM_PROGRAM",
+            "TERM",
+            "TERM_PROGRAM_VERSION",
+            "SHLVL",
+            "ls_colors",
+            "comp_wordbreaks"
+        };
+
+        // Ignore env vars specified in the 'VSO_AGENT_IGNORE' env var
+        private const string EnvIgnore = "VSO_AGENT_IGNORE";
+
+        private Dictionary<string, string> _capsCache;
+
+        public List<Capability> RegularCapabilities
+        {
+            get
+            {
+                return _regularCapabilities;
+            }
+        }
+
+        public List<ToolCapability> ToolCapabilities
+        {
+            get
+            {
+                return _toolCapabilities;
+            }
+        }
+
+        public async Task<Dictionary<string, string>> GetCapabilitiesAsync(string agentName, CancellationToken token)
+        {
+            if (_capsCache != null)
+            {
+                return new Dictionary<string, string>(_capsCache, StringComparer.OrdinalIgnoreCase);
+            }
+
+            var caps = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+
+            GetFilteredEnvironmentVars(caps);
+
+            GetRegularCapabilities(caps, token);
+
+            await GetToolCapabilities(caps, token);
+
+            caps["Agent.Name"] = agentName ?? string.Empty;
+#if OS_LINUX
+            caps["Agent.OS"] = "linux";
+#endif
+
+#if OS_OSX
+            caps["Agent.OS"] = "darwin";
+#endif
+
+#if OS_WINDOWS
+            caps["Agent.OS"] = "Windows_NT";
+#endif
+
+            caps["Agent.ComputerName"] = System.Environment.MachineName ?? string.Empty;
+
+            foreach (var cap in caps)
+            {
+                Trace.Info($"Capability: {cap.Key} Value: {cap.Value}");
+            }
+
+            _capsCache = new Dictionary<string, string>(caps, StringComparer.OrdinalIgnoreCase);
```

Comment:
I'll change the code as in the snippet above. The "_capsCache"  variable is used only in one place: the first 3 lines of GetCapabilitiesAsync:
            if (_capsCache != null)
            {
                return new Dictionary<string, string>(_capsCache, StringComparer.OrdinalIgnoreCase);
            }